### PR TITLE
deps: Bump `rules_foreign_cc` -> 0.10.1

### LINKED ISSUE #30521: Newer release available `rules_foreign_cc`: 0.10.1 (current: 0.9.0)
Package Name: rules_foreign_cc@0.9.0
Current Version: 0.9.0@2022-08-02
Available Version: 0.10.1@2023-10-25
Upstream releases: https://github.com/bazelbuild/rules_foreign_cc/releases

### PR DESCRIPTION
**Pull request by phlax**
Fix #30521 

Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by phlax** [comment]
2nd attempt at this

File: bazel/foreign_cc/BUILD [hunk 1]
```diff
@@ -157,7 +157,7 @@ configure_make(
         "--disable-manual",
         "--disable-shared",
         "--enable-static",
-        "--with-colm=$EXT_BUILD_DEPS/colm",
+        "--with-colm=$$EXT_BUILD_DEPS/colm",
```
Comment (phlax) [inline bazel/foreign_cc/BUILD:160]: not sure what has changed necessitating this - or tbh how it worked previously but it appears to resolve

File: contrib/hyperscan/matching/input_matchers/source/BUILD [hunk 2]
```diff
@@ -30,7 +30,7 @@ envoy_cmake(
         "CMAKE_BUILD_TYPE": "Release",
         "CMAKE_INSTALL_LIBDIR": "lib",
         "FAT_RUNTIME": "on",
-        "RAGEL": "$EXT_BUILD_DEPS/ragel/bin/ragel",
+        "RAGEL": "$$EXT_BUILD_DEPS$$/ragel/bin/ragel",
```
Comment (phlax) [inline contrib/hyperscan/matching/input_matchers/source/BUILD:33]: also not sure why this needs this incantation - but from deciphering the error and looking at upstream it seems to be correct, and works

**Review by phlax — COMMENTED** [review]
there are some more `EXT_BUILD` vars that i havent updated - afaict the invokation required is macro arg specific

**Review by htuch — APPROVED** [review]

**Comment by zirain** [comment]
@phlax seems this broken build on m1.

**Comment by phlax** [comment]
@zirain as this is not tested in ci and i have no way of testing (or even knowing what might have broke) - could you propose a forward fix ?

**Comment by zirain** [comment]
ok, let me try.

**Comment by zirain** [comment]
found this https://gitlab.freedesktop.org/pkg-config/pkg-config/-/issues/81

**Comment by phlax** [comment]
are you certain this PR broke it?

its already known that you cant build with all XCode versions - so wondering if the issue was already present

**Comment by zirain** [comment]
I can build with b27e0b0569 on m1.

**Comment by phlax** [comment]
worth looking here i think https://github.com/bazelbuild/rules_foreign_cc/issues/936

cc @keith

**Comment by phlax** [comment]
@zirain probably its worth opening an actual issue and posting the failure log there - its hard to comment without knowing what the failure is

**Comment by alyssawilk** [comment]
FWIW this broke my build.  Reverting this and the follow-up patches restored and I stopped getting errrors of the form

ERROR: /usr/local/google/home/alyssar/.cache/bazel/_bazel_alyssar/10a2bce46e243a8096192e949a427fad/external/rules_foreign_cc/toolchains/BUILD.bazel:317:15: BootstrapPkgConfig external/rules_foreign_cc/toolchains/pkgconfig [for tool] failed: (Exit 77): bash failed: error executing command (from target @rules_foreign_cc//toolchains:pkgconfig_tool_default) /bin/bash -c bazel-out/k8-opt-exec-A93844C3/bin/external/rules_foreign_cc/toolchains/pkgconfig_tool_default_foreign_cc/wrapper_build_script.sh
....
checking for gcc... clang
checking whether the C compiler works... no

I think it may not play well with https://github.com/envoyproxy/envoy/blob/main/bazel/setup_clang.sh as setting --host_action_env=CC=gcc restored my builds

Just commenting in case anyone else runs into this

**Comment by phlax** [comment]
im a bit confused @alyssawilk 

there was a separate issue which is currently worked around pending upstream fix

> I think it may not play well with https://github.com/envoyproxy/envoy/blob/main/bazel/setup_clang.sh as setting --host_action_env=CC=gcc restored my builds

this is what confuses me - `setup_clang.sh` shouldnt affect a gcc build

are you saying that gcc works but clang doesnt?

if that is the case then my question is what version of llvm are you using - i think our workaround ~fixed macos but i think the real issue is more related to llvm version than platform

(cf https://github.com/bazelbuild/rules_foreign_cc/pull/1187#issuecomment-2035020182)

**Comment by alyssawilk** [comment]
I believe I was using clang so maybe this just broke my clang build.
At the beginning of the week whatever combination of ~/.bazelrc user.bazelrc, aliases etc had my builds complaining
checking for gcc... long/path/to/custom/clang
I assumed something in my env was out of whack so I re-ran our internal "how to get envoy building" which included running bazel/setup_clang.sh 
When I did that my builds switched to 
checking for gcc... clang
So I looked at the arguments in  bazel/setup_clang.sh discovered the magical thing that changed was host_action_env=CC=gcc and adding that to my ~/.bazelrc fixed my builds. so the lua/ioring/etc configure scripts didn't like me using clang instead of gcc for building those external deps. 
I'm happy to play around and see what else I could have changed to keep using clang, but I spent about 2 hours installing every library I could find in any google search result resembling this error and none of them worked.
bumping to llvm16 appears to have gotten me to the next error but rules_foreign_cc is still borked so guess I'll stick with gcc for now.

**Comment by phlax** [comment]
> So I looked at the arguments in bazel/setup_clang.sh discovered the magical thing that changed was host_action_env=CC=gcc

i dont think `setup_clang.sh` ever had `CC=gcc`

> bumping to llvm16 ...

that is known not to work for a bunch of reasons atm

what version of llvm were you using with original issue ?

> ... rules_foreign_cc is still borked 

its neither borked in CI or my local environment - would be good know what is breaking for you (im still thinking the most likely candidate is llvm version - but not sure)

fwiw - i would really like to get rid of `setup_clang.sh` - i think its anti-pattern that creates a load of problems and is unnecessary

**Pull request merged**